cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,234 @@
+
+        LOG.debug("Found cmd.exe at <%s>", cmd)
+        return PurePath(cmd)
+
+    LOG.error("Couldn't find cmd.exe in any mount point.")
+    return None
+
+
+def win_user_profile_dir() -> Optional[PurePath]:
+    """
+    Returns the Windows user profile directory translated as a Linux path
+    accessible inside the current WSL instance.
+    """
+    cmd = cmd_executable()
+    if cmd is None:
+        return None
+
+    # cloud-init runs too early to rely on binfmt to execute Windows binaries.
+    # But we know that `/init` is the interpreter, so we can run it directly.
+    # See /proc/sys/fs/binfmt_misc/WSLInterop[-late]
+    # inside any WSL instance for more details.
+    home, _ = subp.subp(["/init", cmd.as_posix(), "/C", "echo %USERPROFILE%"])
+    home = home.rstrip()
+    if not home:
+        LOG.error("No output from cmd to show the user profile dir.")
+        return None
+
+    return win_path_2_wsl(home.rstrip())
+
+
+def machine_id():
+    """
+    Returns the local machine ID value from /etc/machine-id.
+    """
+    MACHINE_ID_FILE = "/etc/machine-id"
+
+    if util.wait_for_files([MACHINE_ID_FILE], 2.0):
+        LOG.debug("%s file not found", MACHINE_ID_FILE)
+        return None
+
+    return util.load_file(MACHINE_ID_FILE, decode=True)
+
+
+def candidate_user_data_file_names(instance_name) -> List[str]:
+    """
+    Return a list of candidate file names that may contain user-data
+    in some supported format, ordered by precedence.
+    """
+    lsb_rel = util.lsb_release()
+    distribution_id = lsb_rel["id"]
+    release_codename = lsb_rel["codename"]
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, release_codename),
+        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "config.user-data",
+    ]
+
+
+class DataSourceWSL(sources.DataSource):
+    dsname = "WSL"
+
+    def __init__(self, sys_cfg, distro, paths):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths)
+        self._network_config = sources.UNSET
+        self.dsmode = sources.DSMODE_LOCAL
+        self.distro = distro
+        self.instance_name = instance_name()
+
+    def find_user_data_file(self) -> Optional[PurePath]:
+        """
+        Finds the most precendent of the candidate files that may contain
+        user-data, if any, or None otherwise.
+        """
+        profile_dir = win_user_profile_dir()
+        if profile_dir is None:
+            LOG.warning(
+                "Cannot proceed without finding the Windows %USERPROFILE% dir."
+            )
+            return None
+
+        seed_dir = os.path.join(profile_dir, ".cloud-init")
+        if not os.path.isdir(seed_dir):
+            LOG.warning("%s directory doesn't exist.", seed_dir)
```

Comment:
Same, should this be error instead of warning? Or are there conditions that you want to support where the profile_dir does exist, but the seed_dir doesn't exist because there is no applicable user-data for some reason?

Reply:
The other specification we talked about may add more places to look for user-data, but you're right. Let's not anticipate the future here and consider this an error.